presto-main/src/main/resources/webapp/assets/cluster.js

```diff
@@ -0,0 +1,740 @@
+                queuedQueries: addToHistory(clusterState.queuedQueries, this.state.queuedQueries),
+                blockedQueries: addToHistory(clusterState.blockedQueries, this.state.blockedQueries),
+                activeWorkers: addToHistory(clusterState.activeWorkers, this.state.activeWorkers),
+
+                // moving averages
+                runningDrivers: addEwmaToHistory(clusterState.runningDrivers, this.state.runningDrivers),
+                reservedMemory: addEwmaToHistory(clusterState.reservedMemory, this.state.reservedMemory),
+                rowInputRate: addEwmaToHistory(clusterState.rowInputRate, this.state.rowInputRate),
+                byteInputRate: addEwmaToHistory(clusterState.byteInputRate, this.state.byteInputRate),
+                cpuTimeRate: addEwmaToHistory(clusterState.cpuTimeRate, this.state.cpuTimeRate),
+
+                initialized: true,
+            });
+            this.resetTimer();
+        }.bind(this))
+        .error(function() {
+            this.resetTimer();
+        }.bind(this));
+    },
+    componentDidMount: function() {
+        this.refreshLoop();
+    },
+    componentDidUpdate: function() {
+        // prevent multiple calls to setState within the refresh interval from re-rendering sparklines/charts
+        if (this.state.lastRender == null || (Date.now() - this.state.lastRender) >= 1000) {
+            var renderTimestamp = Date.now();
+            $('#running-queries-sparkline').sparkline(this.state.runningQueries, $.extend({}, SPARKLINE_PROPERTIES, {chartRangeMin: 0}));
+            $('#blocked-queries-sparkline').sparkline(this.state.blockedQueries, $.extend({}, SPARKLINE_PROPERTIES, {chartRangeMin: 0}));
+            $('#queued-queries-sparkline').sparkline(this.state.queuedQueries, $.extend({}, SPARKLINE_PROPERTIES, {chartRangeMin: 0}));
+
+            $('#active-workers-sparkline').sparkline(this.state.activeWorkers, $.extend({}, SPARKLINE_PROPERTIES, {chartRangeMin: 0}));
+            $('#running-drivers-sparkline').sparkline(this.state.runningDrivers, $.extend({}, SPARKLINE_PROPERTIES, {numberFormatter: precisionRound}));
+            $('#reserved-memory-sparkline').sparkline(this.state.reservedMemory, $.extend({}, SPARKLINE_PROPERTIES, {numberFormatter: formatDataSizeBytes}));
+
+            $('#row-input-rate-sparkline').sparkline(this.state.rowInputRate, $.extend({}, SPARKLINE_PROPERTIES, {numberFormatter: formatCount}));
+            $('#byte-input-rate-sparkline').sparkline(this.state.byteInputRate, $.extend({}, SPARKLINE_PROPERTIES, {numberFormatter: formatDataSizeBytes}));
+            $('#cpu-time-rate-sparkline').sparkline(this.state.cpuTimeRate, $.extend({}, SPARKLINE_PROPERTIES, {numberFormatter: precisionRound}));
+
+            this.setState({
+                lastRender: renderTimestamp
+            });
+        }
+
+        $('[data-toggle="tooltip"]').tooltip();
+    },
+    render: function() {
```

Comment:
Is it possible to separate renedered templates from classes in React? If so, please move templates to separate files. It would increase code readability.